### DISCUSSION TITLE
Mux Player features not (directly) in Media Chrome (but could be!)

### DISCUSSION THREAD

**Discussion by cjpillsbury** [muxinc/media-chrome#372] [Proposals]
(currently omitting themes and themes-related discussion such as breakpoints/player-size, automatic/well-defined export parts, well defined css vars with well defined semantics for component display, etc. Also currently omitting planned but currently unsupported features such as state management for timed metadata, multiple subtitles/CC, and rendition selection. For either of these, we can can add them to the list if centralization helps)

### Stream Type
* Definition/"Meaning" - Currently defines [several things](https://github.com/muxinc/elements/blob/main/packages/playback-core/src/types.ts#L34-L50) that could be broken up: live vs. on demand, "dvr" vs not, and low latency vs. "standard latency"
* Mux Player/Open Element Purposes
  * (currently) defines [optimized playback-engine config](https://github.com/muxinc/elements/blob/main/packages/playback-core/src/index.ts#L180-L201) _**before**_ loading media content
  * defines [chrome permutations](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/media-theme-mux/media-theme-mux.ts#L176-L231) for the default Mux Player theme
  * overrides standard `mediaplayrequest` behavior to [also seek to live for live content](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/index.ts#L274-L285) (See also: §Seek To Live, below)

### Is Live Edge
* Definition/"Meaning" - A boolean that is true if the playhead is within a computed **_range_** of values that define the live edge of live/DVR content
  * [Implementation of computation fn](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/helpers.ts#L110-L127)
  * [Recalculation conditions](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/index.ts#L287-L296)
* Mux Player/Open Element Purposes
  * Changes the [visualization for Mux Player's custom live indicator/button](https://github.com/muxinc/elements/blob/a2befb53200df08c40a79d7ed84e267cd1279648/packages/mux-player/src/styles.css#L44-L55)
  * [Disables the "play + seek to live" behavior](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/template.ts#L94-L113) of the custom live indicator/button

### Seek To Live
* Definition/"Meaning" - A method that will seek to the live edge of any live/DVR media content.
  * [Implementation of fn](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/helpers.ts#L88-L103)
* Mux Player/Open Elements Purposes
  * Used by Mux Player's custom live indicator/button to [seek to live if not play live edge](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/template.ts#L101-L107)
  * (Hacky workaround) used when `media-play-button` dispatches a `mediaplayrequest` to [also seek to live](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/index.ts#L274-L285) for any live/DVR media content

### Errors
* Definition/"Meaning" - A state+event representing the latest error from Mux Player, [including custom, Mux-Specific errors](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/errors.ts#L7-L163)
* Mux Player/Open Elements Purposes
  * Conditions to [show error dialog to users](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/index.ts#L304-L336)
  * Determines [content shown in error dialog](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/template.ts#L120-L134) (NOTE: `props.dialog` object and values is derived from errors)
  * Provides [custom error info for Mux Data](https://github.com/muxinc/elements/blob/main/packages/playback-core/src/index.ts#L257-L267)

### Shift Subtitle/CC Cues When Controls Are Showing
* Definition/"Meaning" - Tracks when controls are showing (assumed base on [`user-inactive` and `paused`](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/index.ts#L428))
* Mux Player/Open Elements Purposes
  * [Shifts cues up/down](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/index.ts#L410-L461) based on a (currently brittle position-wise) assumed height of the bottom controls when they are showing

### (Error) Dialog
* [Implementation](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/media-chrome/dialog.ts)
* [Usage](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/template.ts#L114-L135)
* State/Architecture - [Open/Closed State Disables UI](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/media-theme-mux/media-theme-mux.ts) (Search for all uses of `isDialogOpen`)

### Live Indicator/Button
* Implementation (N/A, inline button in template)
* [Usage](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/template.ts#L93-L113) (Also see corresponding [attribute-driven CSS](https://github.com/muxinc/elements/blob/main/packages/mux-player/src/styles.css#L40-L55))
* State/Architecture - Various from above used (Seek To Live, Is Live Edge, Stream Type)

**Comment by heff** [comment]
Super helpful! 🎉 

[Finishing up Stream Type here](https://github.com/muxinc/media-chrome/pull/371) (just the live vs. on-demand part of it).

> (currently) defines optimized playback-engine config before loading media content

This feels pretty out of scope of media chrome. The implication of that is a _player_ that wants to configure the _media element_ differently based on stream type or latency desired will need to get that state another way. In Mux Player it looks like we're just using it to configure latency (aside from the other bullet points). Are there any UI impacts when low latency is desired?

> overrides standard mediaplayrequest behavior to [also seek to live for live content](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/index.ts#L274-L285)

It looks like in Mux player we only do this [for the play-button](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/index.ts#L276). Is there any reason this shouldn't also happen on a media element click gesture? i.e. is this a global setting or a play button setting. I'm assuming the former and maybe both.

**Reply by gkatsev** [reply]
> It looks like in Mux player we only do this [for the play-button](https://github.com/muxinc/elements/blob/9226543d5d5e7a74b2e353409d3bd4c0e0993c14/packages/mux-player/src/index.ts#L276). Is there any reason this shouldn't also happen on a media element click gesture? i.e. is this a global setting or a play button setting. I'm assuming the former and maybe both.

That's an oversight we didn't get a chance to address yet.

**Reply by cjpillsbury** [reply]
> That's an oversight we didn't get a chance to address yet.

It's actually because the theme doesn't "know about" `seekToLive()`. We can do this with the live indicator/button bc it's slotted.

**Reply by cjpillsbury** [reply]
> This feels pretty out of scope of media chrome. The implication of that is a player that wants to configure the media element differently based on stream type or latency desired will need to get that state another way. In Mux Player it looks like we're just using it to configure latency (aside from the other bullet points). 

I agree. Calling it out just because it needs to be resolved for a migration.

> Are there any UI impacts when low latency is desired?

Hypothetically, yes? Currently, no. And it would likely just be "on the margins"/for pretty bespoke customization cases, imo, so I don't think we should plan on solving it architecturally in Media Chrome (or Media UI Extensions) unless it comes up.

**Comment by cjpillsbury** [comment]
> Stream Type

Relates to [Stream Type - Proposal](https://github.com/video-dev/media-ui-extensions/issues/3)
Relates to [DVR State - Proposal](https://github.com/video-dev/media-ui-extensions/issues/4)

**Comment by cjpillsbury** [comment]
> Seek To Live

Relates to #382 

**NOTE:** Relies on assumption that a slotted media element's `seekable` will represent the seekable range for live, taking into account things like the implicit or explicit `HOLDBACK` or `PART-HOLDBACK` defined in the HLS spec.

**Comment by cjpillsbury** [comment]
> Is Live Edge

Relates to #382 

**NOTE:** This implementation uses a hard-coded offset value for computing the "live edge window" expediency. A more reliable/accurate implementation would require a Media UI Extension that allowed the slotted media element to advertise what that value should be, based on details of the manifest or playlists in question (e.g. low latency vs. "standard" live, segment durations, etc.)

**Comment by cjpillsbury** [comment]
> Live Indicator/Button

Relates to #382 

**NOTE:** Since the indicator relies on "Is Live Edge" to toggle the live edge indication and enable/disable button functionality, see "Is Live Edge" callout above on its current, simplified implementation.
